File: querydsl-core/src/main/java/com/mysema/query/codegen/BeanSerializer.java
/*
 * Copyright (c) 2010 Mysema Ltd.
 * All rights reserved.
 *
 */
package com.mysema.query.codegen;

import java.io.IOException;
import java.lang.annotation.Annotation;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import com.mysema.codegen.CodeWriter;
import com.mysema.codegen.model.Parameter;
import com.mysema.codegen.model.Types;
import com.mysema.util.BeanUtils;

/**
 * BeanSerializer is a Serializer implementation which serializes EntityType instances into JavaBean classes
 *
 * @author tiwe
 *
 */
public class BeanSerializer implements Serializer{

    private final boolean propertyAnnotations;
    
    private final String javadocSuffix;

    public BeanSerializer() {
        this(true, " is a Querydsl bean type");
    }

    public BeanSerializer(String javadocSuffix) {
        this(true, javadocSuffix);
    }
    
    public BeanSerializer(boolean propertyAnnotations) {
        this(propertyAnnotations, " is a Querydsl bean type");
    }

    public BeanSerializer(boolean propertyAnnotations, String javadocSuffix) {
        this.propertyAnnotations = propertyAnnotations;
        this.javadocSuffix = javadocSuffix;
    }

    @Override
    public void serialize(EntityType model, SerializerConfig serializerConfig, CodeWriter writer) throws IOException {
        String simpleName = model.getSimpleName();

        // package
        if (!model.getPackageName().isEmpty()){
            writer.packageDecl(model.getPackageName());
        }

        // imports
        Set<String> importedClasses = getAnnotationTypes(model);
        if (model.hasLists()){
            importedClasses.add(List.class.getName());
        }
        if (model.hasMaps()){
            importedClasses.add(Map.class.getName());
        }
        writer.importClasses(importedClasses.toArray(new String[importedClasses.size()]));

        // javadoc
        writer.javadoc(simpleName + javadocSuffix);

        // header
        for (Annotation annotation : model.getAnnotations()){
            writer.annotation(annotation);
        }
        writer.beginClass(model);

        // fields
        for (Property property : model.getProperties()){
            if (propertyAnnotations){
                for (Annotation annotation : property.getAnnotations()){
                    writer.annotation(annotation);
                }    
            }            
            writer.privateField(property.getType(), property.getEscapedName());
        }

        // accessors
        for (Property property : model.getProperties()){
            String propertyName = property.getEscapedName();
            // getter
            writer.beginPublicMethod(property.getType(), "get"+BeanUtils.capitalize(propertyName));
            writer.line("return ", propertyName, ";");
            writer.end();
            // setter
            Parameter parameter = new Parameter(propertyName, property.getType());
            writer.beginPublicMethod(Types.VOID, "set"+BeanUtils.capitalize(propertyName), parameter);
            writer.line("this.", propertyName, " = ", propertyName, ";");
            writer.end();
        }

        writer.end();
    }

    private Set<String> getAnnotationTypes(EntityType model) {
        Set<String> imports = new HashSet<String>();
        for (Annotation annotation : model.getAnnotations()){
            imports.add(annotation.annotationType().getName());
        }
        if (propertyAnnotations){
            for (Property property : model.getProperties()){
                for (Annotation annotation : property.getAnnotations()){
                    imports.add(annotation.annotationType().getName());
                }
            }    
        }        
        return imports;
    }

}

File: querydsl-scala/src/main/scala/com/mysema/query/scala/BeanSerializer.scala
package com.mysema.query.scala

import com.mysema.query.codegen._
import com.mysema.codegen.{ CodeWriter, ScalaWriter }
import com.mysema.codegen.model.{ Parameter, Type }
import com.mysema.query

import javax.inject.Inject

import java.util._
import java.io.IOException

import scala.reflect.BeanProperty
import scala.collection.JavaConversions._
import scala.collection.mutable.Set

/**
 * @author tiwe
 *
 */
class ScalaBeanSerializer @Inject() (typeMappings: TypeMappings) extends BeanSerializer(typeMappings) {

  var javadocSuffix = " is a Querydsl bean type"
  
  var javaBeanSupport = false
    
  def writeClass(model: EntityType, writer: ScalaWriter) = {
    // javadoc        
    writer.javadoc(model.getSimpleName + javadocSuffix)
    
    // header
    model.getAnnotations foreach(writer.annotation(_))
    writer.beginClass(model)

    // properties
    model.getProperties foreach { property =>
      property.getAnnotations.foreach(writer.annotation(_))
      if (javaBeanSupport) writer.line("@BeanProperty")
      writer.publicField(property.getType(), property.getEscapedName, "_")
    }

    writer.end()
  }
  
}

/**
 * @author tiwe
 *
 */
class CaseClassSerializer @Inject() (typeMappings: TypeMappings) extends BeanSerializer(typeMappings) {
  
  def javaBeanSupport = false
    
  def writeClass(model: EntityType, writer: ScalaWriter) = {
    model.getAnnotations foreach (writer.annotation(_))
    val parameters = model.getProperties.map(p => new Parameter(p.getEscapedName, p.getType)).toArray
    writer.caseClass(model.getSimpleName, parameters:_*) 
  }

}  
 
/**
 * @author tiwe
 *
 */
abstract class BeanSerializer(typeMappings: TypeMappings) extends Serializer {

  var createCompanionObject = true
  
  def javaBeanSupport: Boolean

  def serialize(model: EntityType, serializerConfig: SerializerConfig, writer: CodeWriter) {
    val simpleName = model.getSimpleName

    // package
    if (!model.getPackageName.isEmpty) writer.packageDecl(model.getPackageName)

    // imports
    val importedClasses = getAnnotationTypes(model)
    if (javaBeanSupport)  importedClasses.add("scala.reflect.BeanProperty")    
    if (model.hasLists()) importedClasses.add(classOf[List[_]].getName)
    if (model.hasMaps())  importedClasses.add(classOf[Map[_, _]].getName)

    writer.importClasses(importedClasses.toArray: _*)
    
    if (createCompanionObject) {
      val queryType = typeMappings.getPathType(model, model, true)
      writeCompanionObject(model, queryType, writer.asInstanceOf[ScalaWriter])  
    } 
    
    writeClass(model, writer.asInstanceOf[ScalaWriter])
  }
  
  def writeClass(model: EntityType, writer: ScalaWriter)
  
  def writeCompanionObject(model: EntityType, queryType: Type, writer: ScalaWriter) = {
    val modelName = writer.getRawName(model)
    val queryTypeName = writer.getRawName(queryType)    
    val variable = model.getUncapSimpleName
    
    writer.beginObject(modelName + " extends "+queryTypeName+"(\""+variable+"\")")
    writer.line("override def as(variable: String) = new ", queryTypeName, "(variable)")
    writer.line("")
    writer.end()
  }  
  
  private def getAnnotationTypes(model: EntityType): Set[String] = {
    val imports = Set() ++ model.getAnnotations.map(_.annotationType.getName);
    // flatMap flattens the results of the map-function.
    // E.g. List(List(1,2,3), List(4,5,6)).flatMap(_.map(_*3)) ends up as List(3, 6, 9, 12, 15, 18).
    imports ++ model.getProperties.flatMap(_.getAnnotations.map(_.annotationType.getName));
  }
  
}


